### ISSUE TITLE
LSP example: is it interpreted in the right way?

### ISSUE THREAD

**Issue by caesarsol** [ryanmcdermott/clean-code-javascript#160]
I wanted to state that in my opinion the "Good" part on the section on **Liskov Substitution Principle** is not a good interpretation.
(I'm really not sure about it, it's a feeling, so forgive me if I'm the one who is interpreting in the wrong way.)

What I think is that the "Bad" and "Good" are too different to be a refactoring example:
 - they are not easily refactorable one in the other, because they are not sharing the inheritance structure
 - it's not clear what problem the code tries to solve
 - the principle is not even applicable in the "Good" example, because there are no instanced objects of type Shape:
   > If Rectangle (S) is a subtype of Shape (T), then objects of type Shape (T) may be replaced with objects of type Rectangle (S)
 - the "Good" example uses a `switch (shape.constructor.name)` which is very not-object-oriented:
   + it hardcodes the class name as a string in the code 
   + it's not open to addition of new classes
   + it violates the [Avoid conditionals](https://github.com/ryanmcdermott/clean-code-javascript#avoid-conditionals) principle, which suggests itself to use classes
   + it violates the [Avoid type-checking](https://github.com/ryanmcdermott/clean-code-javascript#avoid-type-checking-part-1) (I think the Ruby world suggestion of "Tell, don't Ask" would be a good explanation of this principle!)

I'd be glad to discuss and help to build a more correct example, if anyone feels the same on my doubts.

Thanks for the great work!

**Comment by ryanmcdermott** [comment]
Great points, all around. I had trouble getting a Good example to be clearly good. I want to improve the Good case here, and I think we can while still keeping the essence of the example about shapes. Here are my thoughts:

I think we need to remove the `switch` case because that's clearly violating OCP. We want a function that renders large shapes, and it can take squares and rectangles. Maybe what we do is pull out the code for setting length vs. setting height and width, into the declarations part:
```
// Should we set the sides here?
const rectangles = [new Rectangle(), new Rectangle(), new Square()];
renderLargeRectangles(rectangles);
```

**Comment by alemarre** [comment]
In current "bad example" : 
The sentence `If Square is a subtype of Rectangle, then objects of type Rectangle may be replaced with objects of type Square` is actually true.
Square class has `setWidth` and `setHeight` methods, that can be called transparently as if it was a Rectangle.

In current "good example", the same sentence is actually false, you have to check type to avoid calling a method not implemented.

Concerning the comment `// BAD: Will return 25 for Square. Should be 20.` : **why it is bad ?**
You are calling `setWidth` and `setHeight` methods in a row, which, for a Square, link logically to the same method, so obviously the last call take precedence.

I would just inverse the two examples.. and remove the comment `BAD..` :)

**Comment by ryanmcdermott** [comment]
If either of you want to make a PR feel free 😄 I have an idea on how to refactor but would like others' input!

**Comment by ryanmcdermott** [comment]
Closed as per #182
